### ISSUE TITLE
Support for referenced projects

### ISSUE THREAD

**Issue by kentcb** [daveaglick/Buildalyzer#12]
I'm just getting started with Buildalyzer (thanks for your great work!) and have code like this:

```C#
var manager = new AnalyzerManager();
var analyzer = manager.GetProject(initialPath);
var workspace = analyzer.GetWorkspace();
var solution = workspace.CurrentSolution;
```

Whilst this code executes OK, when I examine `solution.Projects` there is only the one project present. This is despite the fact that the project I'm loading references several other projects. They just don't seem to resolve.

I've tried:

* calling `analyzer.Load()`
* calling `analyzer.Compile()`
* enumerating the source files and project references of `analyzer`, which fails because both enumerations are `null`
* explicitly specifying a solution path when creating the `AnalyzerManager` thinking that might help it resolve projects, but it does not

Notes:

* calling `analyzer.Load()` does not return `null`, but calling `analyzer.Compile()` does

To get further with this, I switched to using Buildalyzer source. I found that the [clean](https://github.com/kentcb/Buildalyzer/blob/master/Buildalyzer/ProjectAnalyzer.cs#L149) was returning `false`. To diagnose further I added a logger and got this output:
```
Build started 16/10/2017 12:00:46 PM.

__________________________________________________

Project "C:\Users\Kent\Repository\QUT\Src\UnitTests\UnitTests.csproj" (Clean target(s)):



Target Clean:

    __________________________________________________

    Project "C:\Users\Kent\Repository\QUT\Src\UnitTests\UnitTests.csproj" is building "C:\Users\Kent\Repository\QUT\Src\UnitTests\UnitTests.csproj" (Clean target(s)):



    C:\Program Files\dotnet\sdk\2.0.2\Sdks\Microsoft.NET.Sdk\Sdk\Sdk.props(29,3): error MSB4019: The imported project "C:\Program Files\dotnet\sdk\2.0.2\2.0\Microsoft.Common.props" was not found. Confirm that the path in the <Import> declaration is correct, and that the file exists on disk.



    Done building project "UnitTests.csproj" -- FAILED.
```
I can confirm that the specified `Microsoft.Common.props` file does not exist, but I'm not sure why nor whether it _should_ exist.

**Comment by JosephWoodward** [comment]
Is this not because you're loading a single project? (`var analyzer = manager.GetProject(initialPath);`)

What if you were to pass the [solution when constructing your AnalyzerManager](https://github.com/daveaglick/Buildalyzer/blob/master/src/Buildalyzer/AnalyzerManager.cs#L28)?

**Comment by daveaglick** [comment]
Yep, @JosephWoodward is correct - Buildalyzer doesn't know about solutions (yet). It'll automatically wire up referenced projects when you use the workspaces extension like this:

```csharp
AnalyzerManager manager = new AnalyzerManager();
ProjectAnalyzer analyzer1 = manager.GetProject(@"C:\MyCode\MyProject.csproj");
ProjectAnalyzer analyzer2 = manager.GetProject(@"C:\MyOtherCode\MyOtherProject.csproj");
AdhocWorkspace workspace = new AdhocWorkspace();
analyzer1.AddToWorkspace(workspace);
analyzer2.AddToWorkspace(workspace);
```

In that case, your `workspace.CurrentSolution` will have projects for both `MyProject.csproj` and `MyOtherProject.csproj` and they'll have the correct project references within the Roslyn workspace for each other (but not for any projects that haven't already been loaded by the manager!).

TL;DR: You've got to load each project into the `AnalyzerManager` and add them each to the Roslyn workspace. There's also an extension you can use on the manager: `AdhocWorkspace workspace = manager.GetWorkspace()` that'll load all the projects you've created from the manager into the Roslyn workspace but you've still got to create the projects from the manager first.

I eventually plan to support parsing and loading whole solutions, just haven't gotten there yet.

**Comment by JosephWoodward** [comment]
Question for you @daveaglick,

If my Project A references ProjectB and ProjectC, would loading ProjectA via `manager.GetProject();` also implicitly load ProjectB and C, or would I also need to do that?

**Comment by daveaglick** [comment]
@JosephWoodward Not right now, no. It would only load ProjectA. In fact, the main Buildalyzer library doesn't do anything at all with project references. It's only the Buildalyzer.Workspaces extension that even considers them (and even then, only to wire them up if you manually add them to the workspace).

I've got a couple thoughts here:
* Add an optional `bool` property named something like `LoadProjectReferences` to `AnalyzerManager` that automatically loads referenced projects into the manager when a `ProjectAnalyzer` goes past the `Load()` stage.
* Add a similar argument to the extension methods in the Workspaces extension to do the same - load referenced projects into the manager and then also populate then in the workspace.
* Add support for loading entire solutions in the `AnalyzerManager`.

**Comment by JosephWoodward** [comment]
Sounds like a good idea, happy to take this one on at some point.

**Comment by kentcb** [comment]
> Buildalyzer doesn't know about solutions (
Ah, bummer. Yeah, loading the solution itself was actually the first thing I tried since that's how [my existing Roslyn-only code](https://github.com/kentcb/PCLMock/blob/master/Src/PCLMock.CodeGeneration/Generator.cs#L38-L39) was working.

I will close this issue then, and open another to track solution loading, since that is seemingly what I'll need to progress. My tool can't really know about all projects to load in advance, since it is intended to work against any solution.

**Comment by daveaglick** [comment]
:thumbsup:

Please do open a feature request issue for loading solution files - hope to get that in pretty soon (feel free to grab it @JosephWoodward if you need another PR for Hacktoberfest :smile:)
